ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorPipeline.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+import org.apache.hadoop.hive.ql.metadata.HiveException;
+import java.io.IOException;
+
+/**
+ * Runs different compactions based on the order in which compactors are added.<br>
+ * Mainly used for fall back mechanism for Merge compaction.
+ */
+public class CompactorPipeline {
+
+  private Compactor compactor;
+  private final boolean isMR;
+
+  public CompactorPipeline(Compactor compactor) {
+    this.compactor = compactor;
+    this.isMR = compactor instanceof MRCompactor;
+  }
+
+  CompactorPipeline addCompactor(Compactor newCompactor) {
+    compactor = new FallbackCompactor(compactor, newCompactor);
+    return this;
+  }
+
+  public boolean isMRCompaction() {
+    return isMR;
+  }
+
+  public Boolean execute(CompactorContext input) throws IOException, HiveException, InterruptedException {
+    return compactor.run(input);
+  }
+}
+
+/**
+ * This class defines a way of handling fallback given any number of fallback compactors.<br>
+ * It can encapsulate other fallback compactors within itself.
+ */
+class FallbackCompactor implements Compactor {
```

Comment:
should it be marked as `private static final`?

Reply:
Marked it as `static final` since it is used for testing fallback. Also moved it as an inner class to CompactorPipeline.